test/util/testnode/read.go

```diff
@@ -39,6 +39,63 @@ func ReadBlockchain(ctx context.Context, rpcAddress string) ([]*types.Block, err
 	return ReadBlockHeights(ctx, rpcAddress, 1, status.SyncInfo.LatestBlockHeight)
 }
 
+// ReadBlockchainHeaders retrieves the blockchain headers from height 1 up to
+// latest available height from the node at rpcAddress and returns it.
+// The headers are returned in ascending order (lowest first).
+func ReadBlockchainHeaders(ctx context.Context, rpcAddress string) ([]*types.BlockMeta, error) {
+	client, err := http.New(rpcAddress, "/websocket")
+	if err != nil {
+		return nil, err
+	}
+
+	// fetch the latest height
+	resp, err := client.Status(ctx)
+	if err != nil {
+		return nil, err
+	}
+	maxHeight := resp.SyncInfo.LatestBlockHeight
+
+	blockHeaders := make([]*types.BlockMeta, 0)
+	// fetch headers up to maxHeight
+	lastFetchedHeight := int64(0)
+	println("max height: ", maxHeight)
+	for {
+		// BlockchainInfo may apply a limit on the range of blocks to fetch,
+		// so we need to request them iteratively.
+		// note that block headers returned by BlockchainInfo are in descending
+		// order (highest first).
+		res, err := client.BlockchainInfo(ctx, 1, maxHeight)
+		if err != nil {
+			return nil, err
+		}
+
+		blockHeaders = append(blockHeaders, res.BlockMetas...)
+
+		lastFetchedHeight = res.BlockMetas[len(res.BlockMetas)-1].Header.Height
+
+		// fetch until the first block
+		if lastFetchedHeight <= 1 {
+			break
+		}
+
+		// set the new maxHeight to fetch the next batch of headers
+		maxHeight = lastFetchedHeight - 1
+
+	}
+
+	// reverse the order of headers to be ascending (lowest first).
+	reverseSlice(blockHeaders)
+
+	return blockHeaders, nil
+}
+
+// reverseSlice reverses the order of elements in a slice in place.
+func reverseSlice[T any](s []T) {
+	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
+		s[i], s[j] = s[j], s[i]
+	}
+}
```

Comment:
I guess some tests could be nice to have for this. Some chatGPT table tests I just generated:
```
func TestReverseSlice(t *testing.T) {
	tests := []struct {
		input    interface{}
		expected interface{}
	}{
		{[]int{1, 2, 3, 4, 5}, []int{5, 4, 3, 2, 1}},
		{[]string{"a", "b", "c", "d"}, []string{"d", "c", "b", "a"}},
		{[]int{1, 2}, []int{2, 1}},
		{[]int{1}, []int{1}},
		{[]string{}, []string{}},
	}

	for _, tt := range tests {
		switch v := tt.input.(type) {
		case []int:
			reverseSlice(v)
			if !reflect.DeepEqual(v, tt.expected) {
				t.Errorf("reverseSlice(%v) = %v, want %v", tt.input, v, tt.expected)
			}
		case []string:
			reverseSlice(v)
			if !reflect.DeepEqual(v, tt.expected) {
				t.Errorf("reverseSlice(%v) = %v, want %v", tt.input, v, tt.expected)
			}
		}
	}
}
```

Reply:
Thanks, If you don't mind, I'll add it in a follow up PR.

Reply:
Initially, I intended to address this in a follow-up PR to maintain the current approvals. However, I mistakenly dismissed them. Therefore, I will add the tests in the current PR.

Reply:
Please see https://github.com/celestiaorg/celestia-app/pull/3672/commits/56dc5507a83fd56e013baeaca6dfe87a7a560060